File: App/Services/Teams.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Utility.Strings;

namespace Kandu.Services
{
    public class Teams : Service
    {
        #region "Teams"
        public string List(int orgId)
        {
            if (!IsInOrganization(orgId)) { return AccessDenied(); } //check security
            var list = Query.Teams.GetList(orgId, User.UserId);
            var html = new StringBuilder("{\"teams\":[");
            var i = 0;
            list.ForEach((Query.Models.Team t) =>
            {
                html.Append((i > 0 ? "," : "") + "{\"name\":\"" + t.name + "\", \"description\":\"" + t.description + "\",\"teamId\":\"" + t.teamId + "\"}");
                i++;
            });
            html.Append("]}");
            return html.ToString();
        }

        public string RefreshList(int orgId, bool btnsInFront = false)
        {
            if (!IsInOrganization(orgId)) { return AccessDenied(); } //check security
            var html = "<div class=\"grid-items\">" + Common.Teams.RenderList(this, orgId);
            if (CheckSecurity(orgId, Security.Keys.TeamCanCreate.ToString()))
            {
                var additem = new View("/Views/Teams/add-item.html");
                var addbutton = additem.Render();
                html = (btnsInFront == true ? addbutton : "") + html + (btnsInFront == false ? addbutton : "");
            }
            return html + "</div>";
        }
        #endregion

        #region "Team"

        public string RenderForm(int teamId, int orgId)
        {
            if (!IsInOrganization(orgId)) { return AccessDenied(); } //check security
            var view = new View("/Views/Teams/new-team.html");
            if (teamId != 0)
            {
                //Edit Team form
                var team = Query.Teams.GetTeam(teamId);
                if (!CheckSecurity(team.orgId, new string[] { Security.Keys.TeamCanEditInfo.ToString(), Security.Keys.TeamsFullAccess.ToString() }, Models.Scope.Team, teamId))
                {
                    return AccessDenied();
                }
                view["name"] = team.name;
                view["submit-label"] = "Update Team";
                view["submit-click"] = "S.teams.add.submit('" + teamId + "')";
            }
            else
            {
                //Create Team form
                if (!CheckSecurity(orgId, new string[] { Security.Keys.TeamCanCreate.ToString(), Security.Keys.TeamsFullAccess.ToString() }, Models.Scope.Team, teamId))
                {
                    return AccessDenied();
                }
                view["submit-label"] = "Create Team";
                view["submit-click"] = "S.teams.add.submit()";
            }
            return view.Render();
        }

        public string Create(int orgId, string name, string description = "")
        {
            if (!CheckSecurity(orgId, Security.Keys.TeamCanCreate.ToString())) { return AccessDenied(); } //check security
            Common.Teams.Create(this, orgId, name, description);
            return Success();
        }

        public string Details(int teamId)
        {
            //check security
            var team = Query.Teams.GetTeam(teamId);
            if (!CheckSecurity(team.orgId, new string[] { 
                Security.Keys.TeamCanView.ToString(), Security.Keys.TeamsCanViewAll.ToString(), Security.Keys.TeamsFullAccess.ToString() 
            }, Models.Scope.Team, teamId)) { return AccessDenied(); } 

            //view team details modal
            var canEdit = CheckSecurity(team.orgId, Security.Keys.TeamCanEditInfo.ToString(), Models.Scope.Team, teamId);
            var tabHtml = new StringBuilder();
            var contentHtml = new StringBuilder();
            var view = new View("/Views/Teams/details.html");
            var tab = new View("/Views/Shared/tab.html");
            var html = new StringBuilder();

            //load members tab
            tab["title"] = "Members";
            tab["id"] = "members";
            tab["onclick"] = "S.teams.details.tabs.select('members')";
            tab.Show("selected");
            tabHtml.Append(tab.Render());
            if (team.groupId != null && team.groupId > 0) 
            { 
                html.Append(Common.Teams.RenderMembers(this, teamId));
            }
            if (CheckSecurity(team.orgId, Security.Keys.TeamCanInviteUsers.ToString(), Models.Scope.Team, teamId))
            {
                if (team.groupId != null && team.groupId > 0)
                {
                    var additem = new View("/Views/Members/add-item.html");
                    var addbutton = additem.Render();
                    html.Append(addbutton);
                }
                else
                {
                    html.Append("<div class=\"row pad text-center\"><p>Please set a default security group for your team before adding members.</p></div>");
                }
            }
            contentHtml.Append("<div class=\"content-members grid-items\">" + html.ToString() + "</div>");

            //load invites tab
            tab.Clear();
            tab["title"] = "Invites";
            tab["id"] = "invites";
            tab["onclick"] = "S.teams.details.tabs.select('invites')";
            tabHtml.Append(tab.Render());
            contentHtml.Append("<div class=\"content-invites pad-top\"></div>");

            //load settings tab
            tab.Clear();
            tab["title"] = "Settings";
            tab["id"] = "settings";
            tab["onclick"] = "S.teams.details.tabs.select('settings')";
            tabHtml.Append(tab.Render());
            contentHtml.Append("<div class=\"content-settings pad-top\"></div>");

            view["name"] = team.name;
            view["description"] = team.description;
            view["tabs"] = tabHtml.ToString();
            view["content"] = contentHtml.ToString();
            if (canEdit)
            {
                view.Show("can-edit");
            }
            return view.Render();
        }

        public string Update(int teamId, string name, string description)
        {
            //check security
            var team = Query.Teams.GetTeam(teamId);
            if (!CheckSecurity(team.orgId, new string[] {
                Security.Keys.TeamCanView.ToString(), Security.Keys.TeamCanEditInfo.ToString(), Security.Keys.TeamsFullAccess.ToString()
            }, Models.Scope.Team, teamId)) { return AccessDenied(); }

            //update team info
            Query.Teams.Update(new Query.Models.Team()
            {
                teamId = teamId,
                orgId = team.orgId,
                name = name,
                description = description
            });
            return Success();
        }

        public string RefreshMembers(int orgId, int teamId)
        {
            var team = Query.Teams.GetTeam(teamId);
            if (!CheckSecurity(team.orgId, new string[] {
                Security.Keys.TeamCanView.ToString(), Security.Keys.TeamsCanViewAll.ToString(), Security.Keys.TeamsFullAccess.ToString()
            }, Models.Scope.Team, teamId)) { return AccessDenied(); }
            var html = new StringBuilder("<div class=\"grid-items\">");
            html.Append(Common.Teams.RenderMembers(this, teamId));
            if (CheckSecurity(team.orgId, Security.Keys.TeamCanInviteUsers.ToString(), Models.Scope.Team, teamId))
            {
                if (team.groupId != null && team.groupId > 0)
                {
                    var additem = new View("/Views/Members/add-item.html");
                    var addbutton = additem.Render();
                    html.Append(addbutton);
                }
            }
            return html.ToString() + "</div>";
        }
        #endregion

        #region "Settings"

        public string RefreshSettings(int orgId, int teamId)
        {
            //check security
            var team = Query.Teams.GetTeam(teamId);
            if (!CheckSecurity(team.orgId, new string[] {
                Security.Keys.TeamCanView.ToString(), Security.Keys.TeamsCanViewAll.ToString(), Security.Keys.TeamsFullAccess.ToString()
            }, Models.Scope.Team, teamId)) { return AccessDenied(); }

            //view team settings
            var groups = Query.Security.GetGroups(orgId, User.UserId);
            var view = new View("/Views/Teams/settings.html");
            view.Bind(new { team });
            if(team.groupId == null || team.groupId <= 0)
            {
                view["groups"] = "<option value=\"0\">Select A Security Group...</option>\n";
            }
            view["groups"] += string.Join("\n", groups.Select(a => "<option value=\"" + a.groupId + "\"" + (team.groupId == a.groupId ? " selected" : "") + ">" + a.name + "</option>"));
            return view.Render();
        }

        public string SaveSettings(int orgId, int teamId, Dictionary<string, string> parameters)
        {
            var team = Query.Teams.GetTeam(teamId);
            if (!CheckSecurity(team.orgId, new string[] {
                Security.Keys.TeamCanEditSettings.ToString(), Security.Keys.TeamsFullAccess.ToString()
            }, Models.Scope.Team, teamId)) { return AccessDenied(); }

            var groupId = 0;
            if (parameters.ContainsKey("team_groupid")) int.TryParse(parameters["team_groupid"], out groupId);
            try
            {
                //save org settings
                Query.Teams.UpdateSettings(orgId, teamId, groupId);
            }
            catch (Exception)
            {
                return Error("Error saving team settings");
            }
            try
            {
                //send parameters to all related Partial Views 
                var data = new Dictionary<string, object>()
                {
                    {"orgId", orgId },
                    {"teamId", teamId }
                };
                Common.PartialViews.Save(this, parameters, Vendor.PartialViewKeys.Team_Settings, data);
            }
            catch (Exception)
            {
                return Error("Error saving plugin data for team settings");
            }
            return Success();
        }

        #endregion

        #region "Invite People"
        public string RenderInviteForm(int orgId, int teamId)
        {
            var team = Query.Teams.GetTeam(teamId);
            if (!CheckSecurity(team.orgId, new string[] {
                Security.Keys.TeamCanInviteUsers.ToString(), Security.Keys.TeamsFullAccess.ToString()
            }, Models.Scope.Team, teamId)) { return AccessDenied(); }
            //Invite People to Team form
            var view = new View("/Views/Teams/invite.html");
            if (!CheckSecurity(team.orgId, Security.Keys.TeamCanInviteUsers.ToString(), Models.Scope.Team, teamId) || orgId != team.orgId)
            {
                return AccessDenied();
            }
            view["name"] = team.name;
            view["submit-label"] = "Send Invites";
            view["submit-click"] = "S.teams.invite.submit('" + teamId + "')";
            return view.Render();
        }

        public string RefreshInviteList(int orgId, int teamId, int page = 1, int length = 10, string search = "")
        {
            var team = Query.Teams.GetTeam(teamId);
            if (!CheckSecurity(team.orgId, new string[] {
                Security.Keys.TeamCanInviteUsers.ToString(), Security.Keys.TeamsFullAccess.ToString()
            }, Models.Scope.Team, teamId)) { return AccessDenied(); }
            var m = new Members();
            m.Instantiate(this);
            return m.RefreshList(orgId, page, length, search, true, "Search", teamId, "search by name or email address", "S.teams.invite.select", "S.teams.invite.selectEmail");
        }

        public string RenderMemberSelectedItem(int orgId, int userId)
        {
            if (!IsInOrganization(orgId)) { return AccessDenied(); } //check security
            var view = new View("/Views/Members/list-item-selected.html");
            var member = Query.Users.GetInfo(userId);
            view.Bind(new { member });
            view["remove-onclick"] = "S.teams.invite.removeSelected(event, " + userId + ")";
            return view.Render();
        }

        public string RenderEmailSelectedItem(int orgId, string email)
        {
            if (!IsInOrganization(orgId)) { return AccessDenied(); } //check security
            var view = new View("/Views/Members/use-email-selected.html");
            view["email"] = email;
            view["remove-onclick"] = "S.teams.invite.removeSelectedEmail(event)";
            return view.Render();
        }

        public string InvitePeople(int orgId, int teamId, List<string> people, string message = "")
        {
            var team = Query.Teams.GetTeam(teamId);
            if (!CheckSecurity(team.orgId, new string[] {
                Security.Keys.TeamCanInviteUsers.ToString(), Security.Keys.TeamsFullAccess.ToString()
            }, Models.Scope.Team, teamId)) { return AccessDenied(); }

            //send each person an invite email to your team
            if (team.groupId.HasValue == false || team.groupId <= 0)
            {
                return Error("You must set a default security group for your team before inviting people to join");
            }
            try
            {
                var failed = Common.Invitations.Send(this, people, orgId, Models.Scope.Team, teamId, new string[] { "join" });
                if (failed.Count > 0)
                {
                    return Error(string.Join(",", failed.Select(a => a.name != a.email ? a.name : a.email).ToArray()));
                }
            }
            catch (Exception ex)
            {
                return Error(ex.Message);
            }
            return Success();
        }
        #endregion
    }
}

File: App/Views/Teams/teams.js
﻿S.teams = {
    add: {
        orgId: null,
        popup: null,
        callback: null,

        show: function (id, orgId, callback) {
            S.teams.add.orgId = orgId;
            S.teams.add.callback = callback;
            S.ajax.post('Teams/RenderForm', { teamId: id, orgId: orgId }, (html) => {
                S.teams.add.popup = S.popup.show(!id ? 'Create A New Team' : 'Edit Team', html, {
                    width: 430,
                    onClose: function () {
                        if (callback) { callback(); }
                    }
                });
            });
        },

        hide: function () {
            S.popup.hide(S.teams.add.popup);
        },

        submit: function () {
            var name = $('#teamname').val();
            var description = $('#description').val() || '';
            var msg = $('.popup.show .messages');
            if (name == '' || name == null) {
                S.util.message(msg, 'error', 'Please specify a team name');
                return;
            }
            S.ajax.post('Teams/Create', { orgId: S.teams.add.orgId, name: name, description: description },
                function (data) {
                    S.popup.hide(S.teams.add.popup);
                    if (S.teams.add.callback) {
                        S.teams.add.callback();
                        S.teams.events.broadcast('team-added');
                    }
                },
                function () {
                    S.util.message(msg, 'error', S.message.error.generic);
                    return;
                }
            );
        }
    },

    details: {
        popup: null,
        teamId: null,
        name:'',
        orgId: null,
        callback: null,

        show: function (id, orgId, name, callback) {
            S.teams.details.teamId = id;
            S.teams.details.orgId = orgId;
            S.teams.details.name = name;
            S.teams.details.callback = callback;
            if (typeof name == 'function') { name = name(id, orgId); }
            S.ajax.post('Teams/Details', {teamId: id}, function (result) {
                S.teams.details.popup = S.popup.show('Team ' + name, result, {
                    width: 700,
                    onClose: function () {
                        if (callback) { callback(); }
                    }
                });
                //add events to fields
                $('.team-form input').on('keyup, change', () => {
                    $('.team-form a.apply').removeClass('hide');
                });

                //save button
                $('.team-form a.apply').on('click', S.teams.details.save);

                //set up tabs
                $('.team-details .movable > div').hide();
                $('.team-details .content-members').show();
                $('.team-details .tab-members').on('click', S.teams.members.show);
                $('.team-details .tab-invites').on('click', S.teams.invites.show);
                $('.team-details .tab-settings').on('click', S.teams.settings.show);

                //set up custom scrollbars
                S.scrollbar.add('.team-details .tab-content', { touch: true });

                //update members tab
                S.teams.members.updateEvents();
            });
        },

        editInfo: function () {
            $('.team-details .team-form').removeClass('hide');
            $('.team-details .team-info').hide();
            $('.team-edit-link').hide();
        },

        cancelEdit: function () {
            $('.team-details .team-form').addClass('hide');
            $('.team-details .team-info').show();
            $('.team-edit-link').show();
        },

        hide: function () {
            S.popup.hide(S.teams.details.popup);
        },

        save: function () {
            var form = {
                teamId: S.teams.details.teamId,
                name: $('#teamname').val(),
                description: $('#team_description').val(),
            };
            var msg = $('.popup.show .messages');
            S.ajax.post('Teams/Update', form, function () {
                $('.popup.show .title h5').html('Team ' + form.name);
                if (S.teams.details.callback) {
                    S.teams.details.callback('saved');
                }
                $('.team-details .team-description').html(form.description);
                S.teams.details.cancelEdit();
                S.util.message(msg, null, 'Team updated successfully');
            }, (err) => {
                S.util.message(msg, 'error', err.responseText);
            });
        },

        tabs: {
            select: function (id) {
                $('.team-details .tabs > .tab').removeClass('selected');
                $('.team-details .tabs > .tab-' + id).addClass('selected');
                $('.team-details > .tab-content > .movable > div').hide();
                $('.team-details .content-' + id).show();
                S.popup.resize();
            }
        }
    },

    events: {
        callbacks: [],

        listen: function (callback) {
            S.teams.events.callbacks.push(callback);
        },

        broadcast: function (action, params) {
            var c = S.teams.events.callbacks;
            for (var x = 0; x < c.length; x++) {
                c[x](action, params);
            }
        }
    },

    members: {
        show: function () {
            S.teams.details.tabs.select('members');
            var content = $('.team-details .content-members');
            if (content.html().trim() == '') {
                //load security groups
                S.teams.members.refresh();
            }
        },

        refresh: function () {
            S.ajax.post('Teams/RefreshMembers', { orgId: S.teams.details.orgId, teamId: S.teams.details.teamId, onclick:'S.teams.members.details.show' }, function (result) {
                $('.team-details .content-members').html(result);
                S.teams.members.updateEvents();
            },
            (err) => {

            });
        },

        updateEvents: function () {
            $('.team-details .btn-add-member').on('click', () => {
                S.teams.details.popup.hide();
                S.teams.invite.show(S.teams.details.orgId, S.teams.details.teamId, S.teams.details.name, (total, failed) => {
                    S.teams.details.popup.show();
                    S.util.message('.popup.show .messages', '', total + (total != 1 ? ' people' : ' person') + ' invited' +
                        (failed != null ? ', ' + failed.join(', ') + ' could not be invited' : ''));
                });
            });
            S.popup.resize();
        },

        details: {
            show: function (userId, name) {
                S.teams.details.popup.hide();
                S.user.details.show(userId, S.teams.details.orgId, name, () => {
                    S.teams.details.popup.show();
                });
            }
        }
    },

    invites: {
        show: function () {
            S.teams.details.tabs.select('invites');
            var content = $('.team-details .content-invites');
            if (content.html().trim() == '') {
                //load team settings
                S.teams.invites.refresh();
            }
        },

        refresh: function () {
            S.ajax.post('Teams/RefreshInvites', { orgId: S.teams.details.orgId, teamId: S.teams.details.teamId }, function (result) {
                var content = $('.team-details .content-invites');
                content.html(result);
                var savebtn = $('.team-details .btn-save-invites');
                content.find('select, input').on('keyup, change', () => {
                    savebtn.removeClass('hide');
                });
                S.popup.resize();
            },
            (err) => {

            });
        }
    },

    settings: {
        show: function () {
            S.teams.details.tabs.select('settings');
            var content = $('.team-details .content-settings');
            if (content.html().trim() == '') {
                //load team settings
                S.teams.settings.refresh();
            }
        },

        refresh: function () {
            S.ajax.post('Teams/RefreshSettings', { orgId: S.teams.details.orgId, teamId: S.teams.details.teamId }, function (result) {
                var content = $('.team-details .content-settings');
                content.html(result);
                var savebtn = $('.team-details .btn-save-settings');
                content.find('select, input').on('keyup, change', () => {
                    savebtn.removeClass('hide');
                });
                S.popup.resize();
            },
            (err) => {

            });
        },

        save: function () {
            var savebtn = $('.team-details .btn-save-settings');
            savebtn.addClass('hide');
            var obj = new Object();
            $('.team-details .content-settings').find('select, input, textarea').map((i, a) => {
                a = $(a);
                obj[a.attr('id')] = a.val();
            });
            var data = {
                orgId: S.teams.details.orgId,
                teamId: S.teams.details.teamId,
                parameters: JSON.stringify(obj)
            };
            S.ajax.post('Teams/SaveSettings', data, function (result) {
                S.util.message('.team-details .content-settings .message', null, 'Team settings have been saved');
                S.teams.members.refresh();
            },
                (err) => {
                    S.util.message('.team-details .content-settings .message', 'error', err.responseText);
                });
        }
    },

    invite: {
        orgId: null,
        teamId: null,
        popup: null,
        callback: null,
        cache: null,

        show: function (orgId, teamId, teamName, callback) {
            S.teams.invite.orgId = orgId;
            S.teams.invite.teamId = teamId;
            S.teams.invite.callback = callback;
            S.ajax.post('Teams/RenderInviteForm', { teamId: teamId, orgId: orgId }, (html) => {
                S.teams.invite.popup = S.popup.show('Invite People to Team ' + teamName, html, {
                    width: 700,
                    onClose: function () {
                        if (callback) { callback(); }
                    }
                });
                //initialize the member search form
                S.teams.invite.query(orgId, teamId, 1, 0, '');

                //hide submit button until user selects at least 1 member
                $('.invite-form a.apply').hide();
            });
        },

        hide: function () {
            S.popup.hide(S.teams.invite.popup);
        },

        query: function (orgId, teamId, page, length, search) {
            S.teams.invite.cache = { page: page, length: length, search: search };
            S.ajax.post('Teams/RefreshInviteList', { orgId: orgId, teamId: teamId, page: page, length: length, search: search }, (html) => {
                var inviteform = $('.invite-form .search-users');
                if (length == 0) {
                    inviteform.html(html);

                    //add invited section to form
                    $('.invite-form .members-list').after(
                        '<h5 class="invited-title hide">Selected Invites</h5>' +
                        '<div class="invited-list grid-items hide"></div>');
                    var win = { h: window.innerHeight };
                    $('.invite-form .invited-list, .invite-form .members-list').css({ 'max-height': ((win.h) / 3) + 'px' });

                    //when user searches
                    $('.invite-form form').on('submit', (e) => {
                        e.preventDefault();
                        S.teams.invite.query(orgId, teamId, 1, 20, $('.invite-form #search_members').val());
                        return false;
                    });
                    $('.invite-form .browse-all').attr('onclick', '').on('click', () => {
                        S.teams.invite.query(orgId, teamId, 1, 20, '');
                    });
                } else {
                    $('.invite-form .members-list').html(html);
                    //move results
                    var results = $('.invite-form .members-list .results');
                    var info = $('.invite-form .search-members form > .info');
                    info.html(results[0].outerHTML);
                    results.remove();
                    S.teams.invite.hideSelectedItems();
                }

                //change all item onclick attributes
                inviteform.find('.item > a').each((i, a) => {
                    a.setAttribute('onclick', a.getAttribute('onclick').replace('S.members.details.show(', 'S.teams.invite.select(event, '));
                });
            });
        },

        select: function (id, name) {
            S.ajax.post('Teams/RenderMemberSelectedItem', { orgId: S.teams.invite.orgId, userId: id }, (html) => {
                $('.invited-list').removeClass('hide').prepend(html);
                $('.invited-title').removeClass('hide');
                S.teams.invite.hideSelectedItems();
                S.popup.resize(S.teams.invite.popup);
            });
        },

        selectEmail: function (e, email) {
            //add email to the list of members
            S.ajax.post('Teams/RenderEmailSelectedItem', { orgId: S.teams.invite.orgId, email:email }, (html) => {
                $('.invited-list').removeClass('hide').append(html);
                $('.invited-title').removeClass('hide');
                S.teams.invite.hideSelectedItems();
                S.popup.resize(S.teams.invite.popup);
            });
            $('.invite-form #search_members')[0].value = '';
        },

        hideSelectedItems: function () {
            $('.invite-form .members-list .item').show();
            $('.invited-list .item').each((i, a) => {
                $('.invite-form .members-list .item[data-id="' + a.getAttribute('data-id') + '"]').hide();
            });
            if ($('.invited-list .item').length == 0) {
                $('.invite-form > .buttons a.apply').hide();
            } else {
                $('.invite-form > .buttons a.apply').css({ 'display': 'inline-block' });
            }
        },

        removeSelected: function (e, id) {
            $('.invited-list .item.item-' + id).remove();
            S.teams.invite.hideSelectedItems();
            S.popup.resize(S.teams.invite.popup);
        },

        removeSelectedEmail: function (e) {
            var target = $(e.target);
            if (!target.hasClass('item')) {
                target = target.parents('.item').first();
            }
            target.remove();
            S.teams.invite.hideSelectedItems();
            S.popup.resize(S.teams.invite.popup);
        },

        submit: function () {
            var data = {
                orgId: S.teams.invite.orgId,
                teamId: S.teams.invite.teamId,
                people: $('.invited-list .item').map((i, a) => a.getAttribute('data-id'))
            };
            S.ajax.post('Teams/InvitePeople', data,
                function (response) {
                    S.popup.hide(S.teams.invite.popup);
                    if (S.teams.invite.callback) {
                        S.teams.invite.callback(data.people.length);
                        S.teams.events.broadcast('people-invited');
                    }
                },
                function (err) {
                    S.popup.hide(S.teams.invite.popup);
                    var notinvited = err.responseText.split(',');
                    if (S.teams.invite.callback) {
                        var invited = data.people.length - notinvited.length;
                        S.teams.invite.callback(invited, notinvited);
                        S.teams.events.broadcast('people-invited', { invited: invited, notinvited: notinvited });
                    }
                    return;
                }
            );
        }
    }
}